man/pmaint.1.rst
========
 pmaint
========

----------------------------------------
repository maintenance and modification
----------------------------------------

:Date:   2008-03-09
:Manual section: 1
:Manual group: pkgcore

SYNOPSIS
========

  pmaint [options]

DESCRIPTION
===========

pmaint is the repository mainpulation tool for the pkgcore framework.  Via it
you're able to transfer packages between repositories, regenerate cache data,
'sync' a repository with it's upstream location, and various other repository
specific capabilities.

OPTIONS
=======

.. pkgcore_script_options:: pkgcore.scripts.pmaint.commandline_commands


NEWS
Pkgcore Release Notes
=====================

See ChangeLog for full commit logs; this is summarized/major changes.

* add tar contentsSet rewriting; tarballs sometimes leave out directories,
  and don't always have the fully resolved path- /usr/lib/blah, when
  /usr/lib -> /usr/lib64 *should* be /usr/lib64/blah, but tar doesn't force
  this.  Due to that, can lead to explosions in unpacking- this is now fixed.

* pquery --attr inherited was added; this feature may disappear down the
  line, adding it meanwhile since it's useful for ebuild devs.

* adjust setup.py so that man page installation properly respects --root

* correct a corner case where a package name of 'dev-3D' was flagged as
  invalid.


0.4:

* resolver fixes: vdb loadup wasn't occuring for old style virtuals for
  rdepend blockers, now forces it.  It was possible for a node to be
  considered usable before it's rdepends blockers were leveled- now those
  must be satisfied before being able to dep on the node.

* resolver events cleanup; pmerge now gives far better info as to why a
  choice failed, what it attempted to get around it, etc.

* multiplex trees now discern their frozen state from their subtrees,
  and will execute the repo_op for the leftmost subtree if unfrozen.

* pquery --attr eapi was added.

* ticket 94; package.provided is now supported fully both in profiles,
  and in user profile (/etc/portage/profile).

* ticket 116; ignore empty tarfile exception if the exception explicitly
  states empty header.

* utter corner case compatibility- =dev-util/diffball-1.0-r0 is now the
  same as =dev-util/diffball-1.0 .

* convert FETCHCOMMAND/RESUMECOMMAND support to execute spawn_bash by
  default instead of trying to cut out shell; this kills off the occasional
  incompatibility introduced via portage supplying make.globals.

* FEATURES=sfperms is now a trigger instead of a dyn_preinst hook.
  Faster, cleaner, etc.

* delayed unpacking of binpkgs has been disabled; occasionally can lead to
  quadratic behaviour in contents accessing, and extreme corner case trigger
  breakages.  Will be re-enabled once API has been refactored to remove
  these issues.

* FEATURES=multilib-strict was converted into a trigger.  Tries to
  use the python bindings for file first (merge file[python]), falling
  back to invoking file.  Strongly suggested you have the bindings- fair bit
  faster.  Finally, verification now runs for binpkgs also.

* bug 137; symlink on directory merging failures where pkgcore would wipe
  files it had just installed invalidly.

* correct issue in offset rewriting (was resetting new_offset to '/')-
  should only be api visible, no existing consumers known.

* ebuild env lzma unpack support was broken; fixed (ticket 140).

* Additional debug output for pmerge.

* Further extending PortageFormatter to sanely handle worldfile highlights
  and show repos with both id and location

* Ticket 132: Portage Formatter supports real portage colors now,
  thanks to agaffney for getting the ball rolling

* Masked IUSEs were not treated right in all cases, thanks to agaffney
  for report and help testing

* diefunc tracebacks beautified)


pkgcore 0.3.4:

* IUSEs were filtered, unstated were not respected though breaks with
  current portage tree, so re-enabling.
  Also sanely handle -flag enforcing now and kill hackish code for it.


pkgcore 0.3.3:

* IUSE defaults are respected now, so EAPI=1 implemented

* Write slotted atoms to worldfile as portage supports this now

* Sync up with portage; add support for lzma to unpack- mirror r7991 from
  portage.


pkgcore 0.3.2:

* ticket 190746 from gentoo; basically need to force the perms of first level
  directory of an unpacked $DISTDIR to ensure it's at least readable/writable.
  fixes unpacking of app-misc/screen-4.0.3_p20070403::gentoo-x86 .

* ticket 118; if -u, don't add the node to world set.

* correct a corner case in python implementation of cpv comparison (just
  python, cpy extension handles it correctly); bug 188449 in gentoo, basically
  floats have a limited precision, thus it was possible to get truncation in
  comparison with specially crafted versions.

* handle EOF/IOError on raw_input (for --ask) a bit more gracefully, ticket
  108.

* cd to ${WORKDIR} if ${S} doesn't exist for test/install phases; matches
  change in portage behaviour.

* Now require snakeoil version 0.2 and up- require new capability of 
  AtomicWriteFile, ability to specify uid/gid/perms.  Via that, fixes ticket
  109 (umask leaking through to profile.env).

* the 'glsa' pkgset is now deprecated in favor of 'vuln'; will remain
  through till 0.4 (ticket #106).
 
* ticket 105/96; fix via andkit, basically a bug in einstall lead to
  extra einstall opts getting dropped instead of passed through.

* compatibility fix for lha unpacking for nwere versions of lha.

* emake now invokes ${MAKE:-make}, instead of make- undocumented ebuild
  req, see bug 186598 at bugs.gentoo.org.

* pmerge --verbose is now pmerge -F portage-verbose-formatter

* Stop installing pregen symlink; functionality moved to pmaint regen.

* 'pmerge --domain' was added; basically is a way to specify the domain to
  use, else usees the configuration defined default domain.

* new ebuild trigger to avoid installing files into symlinked dir (get_libdir
  is the friend to fix a common /usr/lib -> /usr/lib64 bug), ticket 119


pkgcore 0.3.1:

* ticket 86; export FILE for portage_conf FETCHCOMMAND/RESUMECOMMAND support,
  convert from spawn_bash to spawn, add some extra error detection

* Correct cleanup of unknown state ebp processors; basically discard them if
  they fail in any way.  Cleanup inherit error msg when under ebd.

* Correct permission issue for vdb virtuals cache.

* ticket 84; rework overlay internals so that sorting order can't accidentally
  expose a version masked by a higher priority repository in an overlay stack.


pkgcore 0.3:

* pregen has moved into pmaint regen.


pkgcore/scripts/pmaint.py
# Copyright: 2005-2007 Brian Harring <ferringb@gmail.com>
# Copyright: 2006 Marien Zwart <marienz@gentoo.org>
# License: GPL2

"""
repository maintainence
"""

__all__ = ('CopyParser', 'DigestParser', 'RegenParser', 'SyncParser')

from pkgcore.util.commandline import convert_to_restrict, OptionParser
from snakeoil.demandload import demandload
demandload(globals(),
    'errno',
    'threading:Event',
    'threading:Thread',
    'Queue:Queue,Empty',
    'time:time,sleep',
    'snakeoil.osutils:pjoin',
    'pkgcore.repository:multiplex',
    'pkgcore.package:mutated',
    'pkgcore.fs:contents,livefs',
    'pkgcore.ebuild:atom,errors,digest',
    'pkgcore.restrictions.boolean:OrRestriction',
)

commandline_commands = {}

def format_seq(seq, formatter=repr):
    if not seq:
        seq = None
    elif len(seq) == 1:
        seq = seq[0]
    else:
        seq = tuple(sorted(str(x) for x in seq))
    return formatter(seq)


class SyncParser(OptionParser):

    def __init__(self, **kwargs):
        OptionParser.__init__(self, description=
            "update a local repository to match its parent", **kwargs)
        self.add_option("--force", action='store_true', default=False,
            help="force an action")

    def check_values(self, values, args):
        values, args = OptionParser.check_values(
            self, values, args)

        if not args:
            values.repos = values.config.repo.keys()
        else:
            for x in args:
                if x not in values.config.repo:
                    self.error("repo %r doesn't exist:\nvalid repos %r" %
                        (x, values.config.repo.keys()))
            values.repos = args
        return values, []

def sync_main(options, out, err):
    """Update a local repositories to match their remote parent"""
    config = options.config
    succeeded, failed = [], []
    seen = set()
    for x in options.repos:
        r = config.repo[x]
        if r in seen:
            continue
        seen.add(r)
        if not r.syncable:
            continue
        out.write("*** syncing %r..." % x)
        if not r.sync(force=options.force):
            out.write("*** failed syncing %r" % x)
            failed.append(x)
        else:
            succeeded.append(x)
            out.write("*** synced %r" % x)
    if len(succeeded) + len(failed) > 1:
        out.write("*** synced %s" % format_seq(sorted(succeeded)))
        if failed:
            err.write("!!! failed sync'ing %s" % format_seq(sorted(failed)))
    if failed:
        return 1
    return 0

commandline_commands['sync'] = (SyncParser, sync_main)


class CopyParser(OptionParser):

    def __init__(self, **kwargs):
        OptionParser.__init__(self, description=
            "copy built pkg(s) into a repository", **kwargs)
        self.add_option("-s", "--source-repo",
            help="copy from just the specified repository; else defaults "
                "to finding any match")
        self.add_option("-t", "--target-repo", default=None,
            help="repository to copy packages into; if specified, "
                "you don't need to specify the target repo as the last arg.  "
                "Mainly useful for xargs invocations")
        self.add_option("--ignore-existing", "-i", default=False,
            action='store_true',
            help="skip existing pkgs, instead of treating it as an overwrite "
            "error")
        self.add_option("--copy-missing", action="store_true", default=False,
            help="Copy packages missing in target repo from source repo")
        self.add_option("--force", action='store_true', default=False,
            help="try and force the copy if the target repository is marked as "
                "immutable")

    def check_values(self, values, args):
        l = len(args)
        if not values.target_repo and l < 2:
            self.error("target_report wasn't specified- specify it either as "
                "the last arguement, or via --target-repo")

        if values.target_repo is not None:
            target_repo = values.target_repo
        else:
            target_repo = args.pop(-1)

        try:
            values.target_repo = values.config.repo[target_repo]
        except KeyError:
            self.error("target repo %r was not found, known repos-\n%s" %
                (target_repo, format_seq(values.config.repo.keys())))

        if values.target_repo.frozen and not values.force:
            self.error("target repo %r is frozen; --force is required to "
                "override this" % target_repo)

        if values.source_repo:
            try:
                values.source_repo = values.config.repo[values.source_repo]
            except KeyError:
                self.error("source repo %r was not found, known repos-\n%s" %
                    (values.source_repo, format_seq(values.config.repo.keys())))
        else:
            values.source_repo = multiplex.tree(*values.config.repos.values())

        values.candidates = []
        if values.copy_missing:
            restrict = OrRestriction(*convert_to_restrict(args))
            for package in values.source_repo.itermatch(restrict):
                if not values.target_repo.match(package.versioned_atom):
                    values.candidates.append(package.versioned_atom)
        else:
            values.candidates = convert_to_restrict(args)

        return values, []


def copy_main(options, out, err):
    """Copy pkgs between repositories."""

    trg_repo = options.target_repo
    src_repo = options.source_repo

    failures = False
    kwds = {'force': options.force}

    for candidate in options.candidates:
        matches = src_repo.match(candidate)
        if not matches:
            err.write("didn't find any matching pkgs for %r" % candidate)
            failures = True
            continue

        for src in matches:
            existing = trg_repo.match(src.versioned_atom)
            args = []
            pkg = src
            if len(existing) > 1:
                err.write(
                    "skipping %r; tried to replace more then one pkg %r..." %
                    (src, format_seq(existing)))
                failures = True
                continue
            elif len(existing) == 1:
                if options.ignore_existing:
                    out.write("skipping %s, since %s exists already" %
                        (src, existing[0]))
                    continue
                out.write("replacing %s with %s... " % (src, existing[0]))
                op = trg_repo.replace
                args = existing
            else:
                out.write("copying %s... " % src)
                op = trg_repo.install

            if src.repo.livefs:
                out.write("forcing regen of contents due to src being livefs..")
                new_contents = contents.contentsSet(mutable=True)
                for fsobj in src.contents:
                    try:
                        new_contents.add(livefs.gen_obj(fsobj.location))
                    except OSError, oe:
                        if oe.errno != errno.ENOENT:
                            err.write("failed accessing fs obj %r; %r\n"
                                "aborting this copy" %
                                (fsobj, oe))
                            failures = True
                            new_contents = None
                            break
                        err.write("warning: dropping fs obj %r since it "
                            "doesn't exist" % fsobj)
                if new_contents is None:
                    continue
                pkg = mutated.MutatedPkg(src, {'contents':new_contents})

            op = op(*(args + [pkg]), **kwds)
            op.finish()

            out.write("completed\n")
    if failures:
        return 1
    return 0

commandline_commands['copy'] = (CopyParser, copy_main)


class RegenParser(OptionParser):

    def __init__(self, **kwargs):
        OptionParser.__init__(
            self, description=__doc__, usage='%prog [options] repo [threads]',
            **kwargs)

    def check_values(self, values, args):
        values, args = OptionParser.check_values(
            self, values, args)
        if not args:
            self.error('Need a repository name.')
        if len(args) > 2:
            self.error('I do not know what to do with more than 2 arguments')

        if len(args) == 2:
            try:
                values.thread_count = int(args[1])
            except ValueError:
                self.error('%r should be an integer' % (args[1],))
            if values.thread_count <= 0:
                self.error('thread count needs to be at least 1')
        else:
            values.thread_count = 1

        try:
            values.repo = values.config.repo[args[0]]
        except KeyError:
            self.error('repo %r was not found! known repos: %s' % (
                    args[0], ', '.join(str(x) for x in values.config.repo)))

        return values, ()


def regen_iter(iterable, err):
    for x in iterable:
        try:
            x.keywords
        except RuntimeError:
            raise
        except Exception, e:
            err.write("caught exception %s for %s" % (e, x))

def reclaim_threads(threads, err):
    for x in threads:
        try:
            x.join()
        except RuntimeError:
            raise
        except Exception, e:
            err.write("caught exception %s reclaiming thread" % (e,))

def regen_main(options, out, err):
    """Regenerate a repository cache."""
    start_time = time()
    # HACK: store this here so we can assign to it from inside def passthru.
    options.count = 0
    if options.thread_count == 1:
        def passthru(iterable):
            for x in iterable:
                options.count += 1
                yield x
        regen_iter(passthru(options.repo), err)
    else:
        queue = Queue(options.thread_count * 2)
        kill = Event()
        kill.clear()
        def iter_queue(kill, qlist, timeout=0.25):
            while not kill.isSet():
                try:
                    yield qlist.get(timeout=timeout)
                except Empty:
                    continue
        regen_threads = [
            Thread(
                target=regen_iter, args=(iter_queue(kill, queue), err))
            for x in xrange(options.thread_count)]
        out.write('starting %d threads' % (options.thread_count,))
        try:
            for x in regen_threads:
                x.start()
            out.write('started')
            # now we feed the queue.
            for pkg in options.repo:
                options.count += 1
                queue.put(pkg)
        except Exception:
            kill.set()
            reclaim_threads(regen_threads, err)
            raise

        # by now, queue is fed. reliable for our uses since the queue
        # is only subtracted from.
        while not queue.empty():
            sleep(.5)
        kill.set()
        reclaim_threads(regen_threads, err)
        assert queue.empty()
    out.write("finished %d nodes in in %.2f seconds" % (options.count,
        time() - start_time))
    return 0

commandline_commands['regen'] = (RegenParser, regen_main)


class DigestParser(OptionParser):

    def __init__(self, **kwargs):
        OptionParser.__init__(
            self, description="generate digests for given atoms", **kwargs)
        self.add_option('-t', '--type', type='choice',
            choices=("manifest1", "manifest2", "both"), default="both",
            help="type of manifest to generate (defaults to both). "
            "valid values are: 'manifest1', 'manifest2', 'both'")

    def check_values(self, values, args):
        values, args = OptionParser.check_values(
            self, values, args)

        if not args:
            self.error('Specify a repo')
        repo = args.pop(0)
        try:
            values.repo = values.config.repo[repo]
        except KeyError:
            self.error("repo %r was not found, known repos-\n%s" %
                (repo, format_seq(values.config.repo.keys())))

        if values.type == "both":
            values.type = ("manifest1", "manifest2")
        else:
            values.type = (values.type,)

        if not args:
            self.error('Specify an atom')
        values.atoms = []
        for arg in args:
            try:
                values.atoms.append(atom.atom(arg))
            except errors.MalformedAtom, e:
                self.error(str(e))

        return values, ()


def digest_main(options, out, err):
    """Write Manifests and digests"""

    for atom in options.atoms:
        pkgs = options.repo.match(atom)
        if not pkgs:
            err.write('No matches for %s\n' % (options.atom,))
            return 1
        for pkg in pkgs:
            if "manifest1" in options.type:
                if options.debug:
                    out.write('Writing digest for %s:' % pkg.cpvstr)
                location = pjoin(pkg.repo.location, pkg.key, "files",
                      "digest-%s-%s" % (pkg.versioned_atom.package,
                                        pkg.versioned_atom.fullver))
                digest.serialize_digest(open(location, 'w'), pkg.fetchables)
            if "manifest2" in options.type:
                if options.debug:
                    out.write('Writing Manifest for %s:' % pkg.cpvstr)
                digest.serialize_manifest("%s/%s" %(pkg.repo.location, pkg.key),
                    pkg.fetchables)

# XXX: harring disabled this for 0.3.
# re-enable it when the bits update manifest.
#commandline_commands['digest'] = (DigestParser, digest_main)
